tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/PathIdManager.scala

```diff
@@ -0,0 +1,182 @@
+        }
+      } yield ()
+    }
+
+  private def deleteOldIdUpdateNewId(
+    pathId: PathId,
+    blobContent: String,
+    idRow: PathId,
+    pathRow: PathId)(implicit session: DBSession): Try[Unit] =
+    if (pathId.timeModified.isAfter(idRow.timeModified) && pathId.timeModified
+          .isAfter(pathRow.timeModified)) {
+      for {
+        _ <- storeAndSendChange(pathId, blobContent)
+        _ <- sendDeleted(pathRow.copy(timeModified = pathId.timeModified))
+        _ <- deletePathId(pathIds, pathId.path)
+        _ <- updatePathId(pathIds, pathId)
+      } yield ()
+    } else Success(())
+
+  private def deleteOldIdInsertNewId(
+    pathId: PathId,
+    blobContent: String,
+    storedPathId: PathId)(implicit session: DBSession): Try[Unit] =
+    if (pathId.timeModified.isAfter(storedPathId.timeModified)) {
+      for {
+        _ <- storeAndSendChange(pathId, blobContent)
+        _ <- sendDeleted(storedPathId.copy(timeModified = pathId.timeModified))
+        _ <- deletePathId(pathIds, pathId.path)
+        _ <- insertPathId(pathIds, pathId)
+      } yield ()
+    } else Success(())
+
+  private def delete(path: String,
+                     timeDeleted: Instant,
+                     maybePathId: Option[PathId],
+                     existingPathId: PathId)(implicit session: DBSession) =
+    if (timeDeleted.isAfter(existingPathId.timeModified)) {
+      for {
+        _ <- sendDeleted(PathId(path, maybePathId.get.id, timeDeleted))
+        _ <- deletePathId(pathIds, path)
+      } yield ()
+    } else Success(())
+
+  private def insert(pathId: PathId, blobContent: String)(
+    implicit session: DBSession) =
+    for {
+      _ <- storeAndSendChange(pathId, blobContent)
+      - <- insertPathId(pathIds, pathId)
+    } yield ()
+
+  private def update(pathId: PathId, blobContent: String, storedPathId: PathId)(
+    implicit session: DBSession) =
+    if (pathId.timeModified.isAfter(storedPathId.timeModified)) {
+      for {
+        _ <- storeAndSendChange(pathId, blobContent)
+        _ <- updatePathId(pathIds, pathId)
+      } yield ()
+    } else Success(())
+  private def sendDeleted(pathId: PathId) =
+    messageSender.sendT[TeiIdMessage](
+      TeiIdDeletedMessage(pathId.id, pathId.timeModified))
+
+  private def storeAndSendChange(pathId: PathId, blobContent: String) =
+    for {
+      stored <- storeTei(pathId, blobContent)
+      _ <- messageSender.sendT[TeiIdMessage](
+        TeiIdChangeMessage(pathId.id, stored.id, pathId.timeModified))
+    } yield ()
+
+  private def storeTei(pathId: PathId, blobContent: String) = {
+    val location = S3ObjectLocation(
+      bucket,
+      s"tei_files/${pathId.id}/${pathId.timeModified.getEpochSecond}.xml")
+    store
+      .put(location)(blobContent)
+      .left
+      .map(error => new RuntimeException(s"Error putting $location: ", error.e))
+      .toTry
+  }
+}
+
+object PathIdManager {
+
+  def selectById(pathIds: PathIdTable, id: String)(
```

Comment:
Minor thing: the use of `pathIds` as a variable name is really throwing me here, because it looks like a `Seq(…)` or `Map(…)` to me. A name like `table` would make this code clearer for me.

Some type hints/comments would also make it easier for me to work out what this functions are doing.

Reply:
sorry only changed the variable names, but didn't add comments. Will do now